compiler/rustc_middle/src/ty/query/plumbing.rs

```diff
@@ -259,52 +256,54 @@ macro_rules! define_queries {
 
         #[allow(nonstandard_style)]
         #[derive(Clone, Debug)]
-        pub enum Query<$tcx> {
-            $($(#[$attr])* $name($($K)*)),*
+        pub struct Query {
+            pub name: &'static str,
+            hash: Fingerprint,
+            description: Cow<'static, str>,
+            span: Option<Span>,
         }
 
-        impl<$tcx> Query<$tcx> {
-            pub fn name(&self) -> &'static str {
-                match *self {
-                    $(Query::$name(_) => stringify!($name),)*
-                }
-            }
-
-            pub fn describe(&self, tcx: TyCtxt<$tcx>) -> Cow<'static, str> {
-                let (r, name) = match *self {
-                    $(Query::$name(key) => {
-                        (queries::$name::describe(tcx, key), stringify!($name))
-                    })*
+        impl Query {
+            $(#[allow(nonstandard_style)] $(#[$attr])*
+            pub fn $name<$tcx>(tcx: TyCtxt<$tcx>, key: $($K)*) -> Self {
+                let kind = dep_graph::DepKind::$name;
+                let name = stringify!($name);
+                let description = queries::$name::describe(tcx, key);
+                let description = if tcx.sess.verbose() {
+                    format!("{} [{}]", description, name).into()
+                } else {
+                    description
                 };
-                if tcx.sess.verbose() {
-                    format!("{} [{}]", r, name).into()
+                let span = if kind == dep_graph::DepKind::def_span {
+                    // The `def_span` query is used to calculate `default_span`,
+                    // so exit to avoid infinite recursion.
+                    None
                 } else {
-                    r
-                }
-            }
+                    Some(key.default_span(tcx))
```

Comment:
This also seems plausibly expensive, FWIW, and presumably not necessary unless we hit an error.